### PR TITLE
Finally use map user to role properly across the application system. 

### PR DESCRIPTION
**Pull request by arnifreyrs**
This method is critical in making application answers and external data read/write-able across the system. I also changed the second parameter for the function so the whole application is accessible.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

**Review by madebynoam — DISMISSED** [review]
The state machine changes look good to me! I like the changes to `mapUserToRole`. The other parts of the PR are out of my expertise, so best that @baering takes a look at those.

**Review by hinriksteinar — DISMISSED** [review]
Looks good to me ( the parts concerning DocumentProviders, other parts not in my expertise ) :+1:

File: apps/application-system/api/src/app/modules/application/utils/tokenUtils.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  ctx: ExecutionContext,
+): string | undefined {
+  const request = ctx.switchToHttp().getRequest()
+  try {
+    const decodedToken = decode(
+      request.headers.authorization?.replace('Bearer ', ''),
+    ) as { nationalId?: string }
+    return decodedToken?.nationalId
+  } catch (e) {
+    return undefined
```
Comment (baering) [inline apps/application-system/api/src/app/modules/application/utils/tokenUtils.ts:14]: `@NationalId() nationalId: string | undefined`

If decodedToken did not contain a nationalId the user is not authorized. Should that not throw an error?

And should the decorator not always be written like this:
`@NationalId() nationalId: string`

So the developer is certain that if the handler in the controller runs, then the user has a valid token.

File: apps/application-system/api/src/app/modules/application/utils/validationUtils.ts [hunk 2]
```diff
@@ -46,15 +46,15 @@ export async function validateIncomingAnswers(
   application: Application,
   newAnswers: FormValue | undefined,
   isStrict = true,
+  nationalId = '',
```
Comment (baering) [inline apps/application-system/api/src/app/modules/application/utils/validationUtils.ts:49]: Same arguments here, will this then not become a required parameter that is not initialised?

File: apps/application-system/api/src/app/modules/application/utils/tokenUtils.ts (hunk 1 above)
Comment (arnifreyrs) [reply apps/application-system/api/src/app/modules/application/utils/tokenUtils.ts:14]: true!

File: libs/api/domains/application/src/lib/application.service.ts [hunk 3]
```diff
@@ -24,81 +24,93 @@ const handleError = (error: any) => {
...
 
-  async deleteAttachment(input: DeleteAttachmentInput) {
+  async deleteAttachment(input: DeleteAttachmentInput, authorization?: string) {
     const { id, ...deleteAttachmentDto } = input
 
     return await this.applicationApi
       .applicationControllerDeleteAttachment({
         id,
         deleteAttachmentDto,
+        authorization,
```
Comment (baering) [inline libs/api/domains/application/src/lib/application.service.ts:113]: `authorization` will become required here

File: libs/application/core/src/types/ApplicationTemplate.ts [hunk 4]
```diff
@@ -25,5 +26,8 @@ export interface ApplicationTemplate<
     states: StatesConfig<TContext, TStateSchema, TEvents> // TODO Extend StatesConfig to completely enforce meta being required attribute
   }
   readonly stateMachineOptions?: Partial<MachineOptions<TContext, TEvents>>
-  mapUserToRole(id: string, state: string): ApplicationRole
+  mapUserToRole(
+    id: string,
+    application: Application,
+  ): ApplicationRole | undefined
```
Comment (baering) [inline libs/application/core/src/types/ApplicationTemplate.ts:32]: Wondering about the return type here. If a user has no role, shouldn't we throw an error?

File: libs/application/core/src/lib/ApplicationTemplateHelper.ts [hunk 5]
```diff
@@ -166,8 +166,11 @@ export class ApplicationTemplateHelper<
     return returnValue
   }
   getWritableAnswersAndExternalData(
-    role: ApplicationRole,
+    role?: ApplicationRole,
   ): ReadWriteValues | undefined {
+    if (!role) {
```
Comment (baering) [inline libs/application/core/src/lib/ApplicationTemplateHelper.ts:171]: Same here with role, like with authentication - should this not just be assumed to be valid and caught elsewhere if not?

**Review by baering — APPROVED** [review]
Nice addition, added some ideas for minor improvements. LGTM 👍

File: libs/application/core/src/types/ApplicationTemplate.ts (hunk 4 above)
Comment (arnifreyrs) [reply libs/application/core/src/types/ApplicationTemplate.ts:32]: i was wondering the same. Not sure what should happen. If there is no role, there will be no form loaded, nor any answers readable or writable any ways. I am a bit afraid of throwing an error, but maybe we can do that later ?

File: apps/application-system/api/src/app/modules/application/e2e/application.spec.ts [hunk 6]
```diff
@@ -71,6 +71,26 @@ describe('Application system API', () => {
     environment.environment = envBefore
   })
 
+  it('should fail when trying to POST when not logged in', async () => {
```
Comment (baering) [inline apps/application-system/api/src/app/modules/application/e2e/application.spec.ts:74]: 💯

**Review by baering — APPROVED** [review]

**Review by hinriksteinar — APPROVED** [review]

**Review by dabbeg — APPROVED** [review]

**Pull request merged**